src/contracts/libraries/GPv2Encoding.sol

```diff
@@ -110,52 +123,79 @@ library GPv2Encoding {
+            // NOTE: Structs are not packed in Solidity, and there is the order
+            // type hash, which is temporarily stored in the slot reserved for
+            // the `owner` field as well as 9 order fields to hash for a total
+            // of `10 * sizeof(uint) = 320` bytes.
+            mstore(order, ORDER_TYPE_HASH)
+            orderDigest := keccak256(order, 320)
         }
 
+        bytes32 signingDigest;
+
         // NOTE: Now compute the digest that was used for signing. This is not
-        // the same as the order digest as it includes a message signature
-        // prefix as well as the domain separator. This is done using a scratch
-        // region of Solidity memory past the last allocation (which is stored
-        // at `0x40` memory address).
+        // the same as the order digest as it is dependant on the signature
+        // scheme being used. In both cases, the signing digest is computed from
+        // a prefix followed by the domain separator and finally the order
+        // the order digest:
+        // - If the order is signed using the EIP-712 sheme, then the prefix is
+        //   the 2-byte value 0x1901,
+        // - If the order is signed using generic message scheme, then the
+        //   prefix is the 28-byte "\x19Ethereum Signed Message\n64" where 64 is
+        //   the length the domain separator and order digest.
         // solhint-disable-next-line no-inline-assembly
         assembly {
-            let scratch := mload(0x40)
-
-            // NOTE: Message prefix is right-padded with 0's so that the first
-            // byte of the message prefix is at `scratch` memory location. 4 0's
-            // are needed as the message prefix is exactly 28 bytes long.
-            // Consequently, the two hashes are stored at 28 and 60 bytes from
-            // the start of the message, and the total length is 92 bytes.
-            mstore(scratch, "\x19Ethereum Signed Message:\n64\x00\x00\x00\x00")
-            mstore(add(scratch, 28), domainSeparator)
-            mstore(add(scratch, 60), orderDigest)
-            signingDigest := keccak256(scratch, 92)
+            // NOTE: Use the region of memory dedicated to the last 4 order
+            // fields to compute the hash in order to prevent further memory
+            // allocations. This is safe as those words have not yet been set
+            // and their values are being stored in stack variables.
+            // Specifically we use:
+            // - `sellTokenIndex` field slot for the prefix (320 byte offset),
+            // - `buyTokenIndex` field slot for the domain separator (352 byte
+            //    offset),
+            // - `executedAmount` field slot for the order digest (384 byte
+            //    offset),
+            mstore(add(order, 352), domainSeparator)
+            mstore(add(order, 384), orderDigest)
+            switch and(v, 0x80)
+                case 0 {
+                    mstore(add(order, 320), 0x1901)
```

Comment:
What the heck is 0x1901? Also should `add(order, 320)` be another stack variable given that it is used frequently as a starting point and from within this context not really clear why.

Reply:
This is the prefix defined in EIP-712 https://github.com/ethereum/EIPs/blob/master/EIPS/eip-712.md#specification

> This encoding is deterministic because the individual components are. The encoding is injective because the three cases always differ in first byte. (RLP_encode(transaction) does not start with \x19.)
> 
> The encoding is compliant with EIP-191. The 'version byte' is fixed to 0x01, the 'version specific data' is the 32-byte domain separator domainSeparator and the 'data to sign' is the 32-byte hashStruct(message).

So the encoding is `"\x19\x01" || domainSeparator || encodeStruct(struct)` where `encodeStruct` is defined as:
`keccak256(abi.encode(structTypeHash, structFields...))`. This is why we compute `orderDigest` to be as it is, so we can reuse it with both signing schemes.